File: flink-scala-core/src/main/scala/com/atguigu/flink/day6/IntervalJoinExample.scala
package com.atguigu.flink.day6

import org.apache.flink.streaming.api.TimeCharacteristic
import org.apache.flink.streaming.api.functions.co.ProcessJoinFunction
import org.apache.flink.streaming.api.scala._
import org.apache.flink.streaming.api.windowing.time.Time
import org.apache.flink.util.Collector

object IntervalJoinExample {

    def main(args: Array[String]): Unit = {
        val env: StreamExecutionEnvironment = StreamExecutionEnvironment.getExecutionEnvironment
        env.setParallelism(1)
        env.setStreamTimeCharacteristic(TimeCharacteristic.EventTime)

        val stream1: KeyedStream[(String, Long, String), String] = env.fromElements(
            ("user_1", 10 * 60 * 1000L, "click")
        ).assignAscendingTimestamps(_._2)
                .keyBy(_._1)

        val stream2: KeyedStream[(String, Long, String), String] = env.fromElements(
            ("user_1", 5 * 60 * 1000L, "browse"),
            ("user_1", 6 * 60 * 1000L, "browse")
        ).assignAscendingTimestamps(_._2)
                .keyBy(_._1)

      /*  stream1.intervalJoin(stream2)
                        .between(Time.minutes(-10),Time.minutes(0))
                        .process(new ProcessJoinFunction[(String,Long,String),(String,Long,String),String] {
                            override def processElement(left: (String, Long, String), right: (String, Long, String), ctx: ProcessJoinFunction[(String, Long, String), (String, Long, String), String]#Context, out: Collector[String]) = {
                                out.collect(left + " = > " +right)
                            }
                        }).print()*/
        stream2.intervalJoin(stream1)
                        .between(Time.minutes(0),Time.minutes(10))
                        .process(new ProcessJoinFunction[(String,Long,String),(String,Long,String),String] {
                            override def processElement(left: (String, Long, String), right: (String, Long, String), ctx: ProcessJoinFunction[(String, Long, String), (String, Long, String), String]#Context, out: Collector[String]) = {
                                out.collect(left + " = > " +right)
                            }
                        }).print()
        env.execute()
    }

}

File: flink-java-core/src/main/java/com/atguigu/flink/day6/IntervalJoinExample.java
package com.atguigu.flink.day6;

import org.apache.flink.api.common.eventtime.SerializableTimestampAssigner;
import org.apache.flink.api.common.eventtime.WatermarkStrategy;
import org.apache.flink.api.java.tuple.Tuple3;
import org.apache.flink.streaming.api.TimeCharacteristic;
import org.apache.flink.streaming.api.datastream.KeyedStream;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.streaming.api.functions.co.ProcessJoinFunction;
import org.apache.flink.streaming.api.windowing.time.Time;
import org.apache.flink.util.Collector;

/**
 * @author ：hmz
 * @date ：Created in 2020/9/16 13:59
 */
public class IntervalJoinExample {
    public static void main(String[] args) throws Exception{

        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
        env.setParallelism(1);
        env.setStreamTimeCharacteristic(TimeCharacteristic.EventTime);

        KeyedStream<Tuple3<String, Long, String>, String> stream1 = env.fromElements(
                Tuple3.of("user_1", 10 * 60 * 1000L, "click")
        ).assignTimestampsAndWatermarks(
                WatermarkStrategy.<Tuple3<String, Long, String>>forMonotonousTimestamps()
                        .withTimestampAssigner(new SerializableTimestampAssigner<Tuple3<String, Long, String>>() {
                            @Override
                            public long extractTimestamp(Tuple3<String, Long, String> element, long recordTimestamp) {
                                return element.f1;
                            }
                        })
        ).keyBy(r -> r.f0);
        KeyedStream<Tuple3<String, Long, String>, String> stream2 = env.fromElements(
                Tuple3.of("user_1", 5 * 60 * 1000L, "browse"),
                Tuple3.of("user_1", 6 * 60 * 1000L, "browse")
        ).assignTimestampsAndWatermarks(
                WatermarkStrategy.<Tuple3<String, Long, String>>forMonotonousTimestamps()
                        .withTimestampAssigner(new SerializableTimestampAssigner<Tuple3<String, Long, String>>() {
                            @Override
                            public long extractTimestamp(Tuple3<String, Long, String> element, long recordTimestamp) {
                                return element.f1;
                            }
                        })
        ).keyBy(r -> r.f0);

        stream1.intervalJoin(stream2)
                .between(Time.minutes(-10),Time.minutes(0))
                .process(new ProcessJoinFunction<Tuple3<String, Long, String>, Tuple3<String, Long, String>, String>() {
                    @Override
                    public void processElement(Tuple3<String, Long, String> left, Tuple3<String, Long, String> right, Context ctx, Collector<String> out) throws Exception {
                        out.collect(left +"=>" + right);
                    }
                })
                .print();

        env.execute();

    }

}
